tasks/main.yml

```diff
@@ -1,25 +1,115 @@
 # SPDX-License-Identifier: MIT
 ---
-- name: Set platform/version specific variables
-  include_tasks: tasks/set_vars.yml
-
-# Examples of some tasks:
 - name: Ensure required packages are installed
   package:
-    name: "{{ __template_packages }}"
+    name: "{{ __systemd_packages }}"
     state: present
 
-- name: Ensure required services are enabled and started
-  service:
+- name: Deploy unit files
+  copy:
+    src: "{{ item }}"
+    dest: "{{ __admin_units_dir }}/{{ item }}"
+    owner: root
+    group: root
+    mode: '0644'
+  loop: "{{ systemd_unit_files }}"
+  register: __systemd_unit_files_result
+
+- name: Deploy unit file templates
+  template:
+    src: "{{ item }}.j2"
+    dest: "{{ __admin_units_dir }}/{{ item }}"
+    owner: root
+    group: root
+    mode: '0644'
+  loop: "{{ systemd_unit_file_templates }}"
+  register: __systemd_unit_file_templates_result
+
+- name: Create drop-in directories
+  file:
+    path: "{{ __admin_units_dir }}/{{ '.'.join(item.split('.')[:-1]) }}.d/"
+    state: directory
+    owner: root
+    group: root
+    mode: '0755'
+  loop: "{{ systemd_dropins }}"
+
+- name: Deploy systemd drop in config
+  template:
+    src: "{{ item }}.j2"
+    dest: "{{ __admin_units_dir }}/\
+     {{ '.'.join(item.split('.')[:-1]) }}.d/\
+     99-override.conf"
+    owner: root
+    group: root
+    mode: '0644'
+  loop: "{{ systemd_dropins }}"
+  register: __systemd_dropins_result
+
+- name: Reload systemd
+  systemd:
+    daemon_reload: true
+  when: >
+    __systemd_unit_files_result is changed or
+    __systemd_unit_file_templates_result is changed or
+    __systemd_dropins_result is changed
+
+- name: Ensure required units are started
+  systemd:
     name: "{{ item }}"
     state: started
+  when: systemd_started_units is defined
```

Comment:
```suggestion
```
anything defined in defaults/main.yml will always be defined - the `loop` with an empty list is a no-op, so you don't need a conditional at all

Same below for the other tasks - you don't need the `when: xxx is defined` if xxx is defined in defaults/main.yml